R/download-file.R

```diff
@@ -0,0 +1,155 @@
+possibly <- function(.f, otherwise, quiet = TRUE) {
+  force(otherwise)
+  function(...) {
+    tryCatch(
+      .f(...),
+      error = function(e) {
+        if (!quiet)
+          message("Error: ", e$message)
+        otherwise
+      },
+      interrupt = function(e) {
+        stop("Terminated by user", call. = FALSE)
+      }
+    )
+  }
+}
+
+safe_GET <- possibly(GET, NULL, quiet = TRUE)
+
+#' Download file from the Internet (cache-aware)
+#'
+#' This is an alternative to [utils::download.file()] and a convenience wrapper for
+#' [GET()] + [httr::write_disk()] to perform file downloads.
+#'
+#' Since this function uses [GET()], callers can pass in `httr` configuration
+#' options to customize the behaviour of the download process (e.g. specify a `User-Agent` via
+#' [user_agent()], set proxy config via [use_proxy()], etc.).
+#'
+#' The function is also "cache-aware" in the sense that you deliberately have to specify
+#' `overwrite = TRUE` to force a re-download. This has the potential to save bandwidth
+#' of both the caller and the site hosting files for download.
+#'
+#' @note While this function supports specifying multiple URLs and download paths it
+#'       does not perform concurrent downloads.
+#' @param url the url(s) of the file to retrieve. If multiple URLs are provided then the same
+#'        number of `path`s must also be provided.
+#' @param path Path(s) to save content to. If more than one `path` is specified then the same
+#'        number of `url`s must also be provided. THis parameter will be [path.expand()]ed.
+#' @param overwrite	Will only overwrite existing path if `TRUE`.
+#' @param ... passed on to [GET()]
+#' @return a data frame containing the `url`(s), `path`(s), cache status, and HTTP status code(s).
+#'         If there was an error downloading a file the path, status code, and HTTP status
+#'         columns will be `NA`. If the file was now re-downloaded the status code will be 399
+#' @seealso [GET()]; [write_disk()]
+#' @export
+#' @examples
+#' tmp1 <- tempfile()
+#' tmp2 <- tempfile()
+#' tmp3 <- tempfile()
+#'
+#' download_file("https://google.com", tmp1) # downloads fine
+#' download_file("https://google.com", tmp1) # doesn't re-download since it's cached
+#' download_file("https://google.com", tmp1, overwrite = TRUE) # re-downloads (overwrites file)
+#' download_file("https://google.com", tmp2) # re-downloads (new file)
+#' download_file("badurl", tmp3) # handles major errors gracefully
+#'
+#' # multi-file example with no-caching
+#' download_file(
+#'   c(rep("https://google.com", 2), "badurl"),
+#'   c(tmp1, tmp2, tmp3),
+#'   overwrite = TRUE
+#' )
+#'
+#' # multi-file example with caching
+#' download_file(
+#'   c(rep("https://google.com", 2), "badurl"),
+#'   c(tmp1, tmp2, tmp3),
+#'   overwrite = FALSE
+#' )
+download_file <- function(url, path, overwrite = FALSE, ...) {
+
+  url <- as.character(url)
+  path <- as.character(path)
```

Comment:
This makes me think of something I do in `usethis::use_course()` and recently decided to not present as a PR into curl (https://github.com/jeroen/curl/issues/187): trying to determine local file name from the Content-Disposition header (and falling back to base name of the URL).

@hrbrmstr Do you have any thoughts on that? Maybe the curl issue is the best place to put them btw.

Reply:
That's a super gd question. I have a similar idiom setup for some light crawls we do but hadn't thought to make it generic somewhere.

I wonder if adding a "helper" to/for `write_disk()` might be a way to go (not sure `curl` is the right place for it).

Something like (seriously off the top of my head) `write_disk <- function(path, infer = FALSE, overwrite = FALSE)`

where if `infer` were `TRUE` it'd have the behaviour you posited:

- look for `Content-Disposition` and use that if found (potentially sanitizing it since it's theoretically untrusted content)
- use last part of URL if ^^ does not exist, BUT
- use a generated UUID-ish name if ^^ is a `/`

AND store it in the directory pointed to in `path`.

Reply:
What you describe it basically exactly what I do in usethis (Content-Disposition > last bit of URL > random name + sanitization):

https://github.com/r-lib/usethis/blob/master/R/course.R

Reply:
👍 It'd be pretty straightforward for me to add this to `write_disk()` (on holiday in 2 days so wldn't be until after that unless it seriously downpours in Acadia :-) if folks deem that a worthy add.